httputil/retry.go

```diff
@@ -139,6 +147,20 @@ func RetryRequest(endpoint string, doRequest func() (*http.Response, error), rea
 			if ShouldRetryError(attempt, err) {
 				continue
 			}
+
+			// on 16.04 the error below is "temporary", however on 18.04 it is not, so we can't relay on Temporary() flag and instead need
+			// to inspect the inner errors; note, that means the error was actually retried on 16.04 (due to the ShouldRetryError logic).
+			if urlErr, ok := err.(*url.Error); ok {
+				if netopErr, ok := urlErr.Err.(*net.OpError); ok {
+					if netopErr.Op == "dial" {
+						if dnsError, ok := netopErr.Err.(*net.DNSError); ok {
+							if strings.Contains(dnsError.Err, "connect: network is unreachable") || strings.Contains(dnsError.Err, "Temporary failure in name resolution") {
```

Comment:
Do you think those messages could become localised?

Reply:
Huh, that's a good question! That would potentially affect existing retry code (similiar check there). I'll verify

Reply:
Uh, "connect: network is unreachable" appears to be defined in golang sources and is not translated. However "Temporary failure in name resolution" appears in glibc sources and has translations. So I think we have a potential problem also in similiar case in `ShouldRetryError(` logic 😭

Reply:
Ok, I think we're fine after all, I tested this with a simple piece of go code under PL locale and the "Temporary..." error message comes untranslated to go, whereas I'm getting the translated one only when using cli tools (e.g. ping).

Reply:
I looked at this a bit and I think with careful unpacking of the errors we can avoid some of the string matches. Here is what I have in mind: http://paste.ubuntu.com/p/GdmN4vQWdr/

The first part will check for errors from e.g. connect when the network is unreachable or down, it looks at the errno from the syscall. The second part checks DNS, the error we are looking for is EAI_AGAIN, however this is not directly exposed anywhere which really sucks. I created https://github.com/golang/go/pull/31676 to make things a bit better but that won't help in the short term :/

Reply:
Thanks for this patch, I look this structure with two helpers a lot, and it's also more future-proof in case Go improves in that area. Applied, with just small tweaks due to different (and unfortunate) behavior on 16.04.